packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2850,24 +2839,63 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         }
     }
 
+    private async waitForDeltaManagerToCatchup(
+        latestSnapshotRefSeq: number,
+        summaryLogger: ITelemetryLogger,
+    ): Promise<void> {
+        if (latestSnapshotRefSeq > this.deltaManager.lastSequenceNumber) {
+            // We need to catch up to the latest summary's reference sequence number before proceeding.
+            await PerformanceEvent.timedExecAsync(
+                summaryLogger,
+                {
+                    eventName: "WaitingForSeq",
+                    lastSequenceNumber: this.deltaManager.lastSequenceNumber,
+                    targetSequenceNumber: latestSnapshotRefSeq,
+                    lastKnownSeqNumber: this.deltaManager.lastKnownSeqNumber,
+                },
+                async () => waitForSeq(this.deltaManager, latestSnapshotRefSeq),
+                { start: true, end: true, cancel: "error" }, // definitely want start event
+            );
+        }
+    }
+
     /** Implementation of ISummarizerInternalsProvider.refreshLatestSummaryAck */
     public async refreshLatestSummaryAck(options: IRefreshSummaryAckOptions) {
         const { proposalHandle, ackHandle, summaryRefSeq, summaryLogger } = options;
         const readAndParseBlob = async <T>(id: string) => readAndParse<T>(this.storage, id);
         // The call to fetch the snapshot is very expensive and not always needed.
         // It should only be done by the summarizerNode, if required.
+        // When fetching from storage we will always get the latest version and do not use the ackHandle.
```

Comment:
I still think we should handle scenarios where this can be called when a refresh is already in progress. We should not rely on the caller doing the right thing.
At least, add asserts or telemetry when this happens so we know someone screwed up. The telemetry will tell us if our assumption is wrong.

Reply:
Don't we have a lock in place for this not to happen ?

Reply:
yeah but the lock is in summarizer classes. What if there is a bug and it ends up calling it multiple times. Adding an assert or log will help identify these scenarios rather than trying to figure it from other logs which has turned out to be quite cumbersome.